From ebe29fd9cc0d8eaed4ffb652c128f25ca0e78c0f Mon Sep 17 00:00:00 2001
From: hienhuynh <hien.huynh.px@renesas.com>
Date: Thu, 29 Apr 2021 13:33:24 +0700
Subject: [PATCH 3/3] core: arm: plat-rzg: Add ECC mode checking for shared
 memory region

When we use ECC setting, the memory and size
of DDR we can use may be changed. So that we must modify them to
compatible with the system(and un-trusted OS).

Otherwise, these macros are used in our platform source only, so that
move to platform private header to avoid warning of undefined macro
value when other source use these pre-define.

Signed-off-by: Huynh Thanh Hung <hung.huynh.wz@renesas.com>
Signed-off-by: hienhuynh <hien.huynh.px@renesas.com>
---
 core/arch/arm/plat-rzg/conf.mk           |  2 +
 core/arch/arm/plat-rzg/main.c            |  2 +
 core/arch/arm/plat-rzg/platform_config.h | 29 ----------
 core/arch/arm/plat-rzg/rzg_common.h      | 94 ++++++++++++++++++++++++++++++++
 core/arch/arm/plat-rzg/sub.mk            |  2 +-
 5 files changed, 99 insertions(+), 30 deletions(-)
 create mode 100644 core/arch/arm/plat-rzg/rzg_common.h

diff --git a/core/arch/arm/plat-rzg/conf.mk b/core/arch/arm/plat-rzg/conf.mk
index e3fa853..cd3d04a 100644
--- a/core/arch/arm/plat-rzg/conf.mk
+++ b/core/arch/arm/plat-rzg/conf.mk
@@ -44,3 +44,5 @@ $(call force,CFG_ARM32_core,y)
 endif
 
 CFG_DT ?= y
+RZG_DRAM_ECC ?= 0
+RZG_ECC_FULL ?= 0
diff --git a/core/arch/arm/plat-rzg/main.c b/core/arch/arm/plat-rzg/main.c
index 6a1021b..0409ef8 100644
--- a/core/arch/arm/plat-rzg/main.c
+++ b/core/arch/arm/plat-rzg/main.c
@@ -11,6 +11,8 @@
 #include <mm/core_memprot.h>
 #include <platform_config.h>
 
+#include "rzg_common.h"
+
 register_phys_mem_pgdir(MEM_AREA_IO_SEC, DEVICE0_BASE, DEVICE0_SIZE);
 register_phys_mem_pgdir(MEM_AREA_IO_SEC, CONSOLE_UART_BASE, SCIF_REG_SIZE);
 register_phys_mem_pgdir(MEM_AREA_IO_SEC, GICD_BASE, GIC_DIST_REG_SIZE);
diff --git a/core/arch/arm/plat-rzg/platform_config.h b/core/arch/arm/plat-rzg/platform_config.h
index 0638b53..0a0a2b4 100644
--- a/core/arch/arm/plat-rzg/platform_config.h
+++ b/core/arch/arm/plat-rzg/platform_config.h
@@ -21,35 +21,6 @@
 
 #define CONSOLE_UART_BASE	0xE6E88000
 
-#if defined(PLATFORM_FLAVOR_ek874)
-#define NSEC_DDR_0_BASE		0x47E00000U
-#define NSEC_DDR_0_SIZE		0x78200000
-
-#elif defined(PLATFORM_FLAVOR_hihope_rzg2h)
-
-#define NSEC_DDR_0_BASE		0x47E00000U
-#define NSEC_DDR_0_SIZE		0x78200000
-#define NSEC_DDR_1_BASE		0x500000000U
-#define NSEC_DDR_1_SIZE		0x80000000
-
-#elif defined(PLATFORM_FLAVOR_hihope_rzg2m)
-
-#define NSEC_DDR_0_BASE		0x47E00000U
-#define NSEC_DDR_0_SIZE		0x78200000
-#define NSEC_DDR_1_BASE		0x600000000U
-#define NSEC_DDR_1_SIZE		0x80000000
-
-#elif defined(PLATFORM_FLAVOR_hihope_rzg2n)
-
-#define NSEC_DDR_0_BASE		0x47E00000U
-#define NSEC_DDR_0_SIZE		0x78200000
-#define NSEC_DDR_1_BASE		0x480000000U
-#define NSEC_DDR_1_SIZE		0x80000000
-
-#else
-#error "Unknown platform flavor"
-#endif
-
 #define TEE_SHMEM_START		(TZDRAM_BASE + TZDRAM_SIZE)
 #define TEE_SHMEM_SIZE		0x100000
 
diff --git a/core/arch/arm/plat-rzg/rzg_common.h b/core/arch/arm/plat-rzg/rzg_common.h
new file mode 100644
index 0000000..7b41df9
--- /dev/null
+++ b/core/arch/arm/plat-rzg/rzg_common.h
@@ -0,0 +1,94 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2020, Renesas Electronics Corporation
+ */
+
+#ifndef RZG_COMMON_H
+#define RZG_COMMON_H
+
+#include <string.h>
+#include <kernel/misc.h>
+#include <kernel/panic.h>
+#include <arm.h>
+
+#if defined(PLATFORM_FLAVOR_ek874)
+#if (RZG_DRAM_ECC == 1 && RZG_ECC_FULL == 2)
+/* Apply for ECC Full Single setting */
+#define NSEC_DDR_0_BASE		0x47E00000U
+#define NSEC_DDR_0_SIZE		0x3C200000
+#else //(RZG_DRAM_ECC == 1 && RZG_ECC_FULL == 2)
+/* Apply for non-ECC or Partial ECC setting */
+#define NSEC_DDR_0_BASE		0x47E00000U
+#define NSEC_DDR_0_SIZE		0x78200000
+#endif //(RZG_DRAM_ECC == 1 && RZG_ECC_FULL == 2)
+
+#elif defined(PLATFORM_FLAVOR_hihope_rzg2h)
+#if (RZG_DRAM_ECC == 1)
+#if (RZG_ECC_FULL == 1)
+/* Apply for ECC Full Dual setting */
+#define NSEC_DDR_0_BASE		0x47E00000U
+#define NSEC_DDR_0_SIZE		0x78200000
+#elif (RZG_ECC_FULL == 2)
+/* Apply for ECC Full Single setting */
+#define NSEC_DDR_0_BASE		0x47E00000U
+#define NSEC_DDR_0_SIZE		0x4C200000
+#define NSEC_DDR_1_BASE		0x600000000U
+#define NSEC_DDR_1_SIZE		0x40000000
+#else //(RZG_ECC_FULL)
+/* Apply for ECC Partial setting */
+/* Define full of DDR0/DDR1 memory map*/
+#define NSEC_DDR_0_BASE		0x47E00000U
+#define NSEC_DDR_0_SIZE		0x78200000
+#define NSEC_DDR_1_BASE		0x600000000U
+#define NSEC_DDR_1_SIZE		0x80000000
+#endif
+#else //(RZG_DRAM_ECC == 1)
+/* Define original memory*/
+#define NSEC_DDR_0_BASE		0x47E00000U
+#define NSEC_DDR_0_SIZE		0x78200000
+#define NSEC_DDR_1_BASE		0x500000000U
+#define NSEC_DDR_1_SIZE		0x80000000
+#endif //(RZG_DRAM_ECC == 1
+
+#elif defined(PLATFORM_FLAVOR_hihope_rzg2m)
+
+#if (RZG_DRAM_ECC == 1 && RZG_ECC_FULL == 1)
+/* Apply for ECC Full Dual setting */
+#define NSEC_DDR_0_BASE		0x47E00000U
+#define NSEC_DDR_0_SIZE		0x78200000
+#elif (RZG_DRAM_ECC == 1 && RZG_ECC_FULL == 2)
+/* Apply for ECC Full Single setting */
+#define NSEC_DDR_0_BASE		0x47E00000U
+#define NSEC_DDR_0_SIZE		0x4C200000
+#define NSEC_DDR_1_BASE		0x600000000U
+#define NSEC_DDR_1_SIZE		0x40000000
+#else
+/* Apply for non-ECC or Partial ECC setting */
+/* Define original memory*/
+#define NSEC_DDR_0_BASE		0x47E00000U
+#define NSEC_DDR_0_SIZE		0x78200000
+#define NSEC_DDR_1_BASE		0x600000000U
+#define NSEC_DDR_1_SIZE		0x80000000
+#endif
+
+#elif defined(PLATFORM_FLAVOR_hihope_rzg2n)
+
+#if (RZG_DRAM_ECC == 1 && RZG_ECC_FULL == 2)
+/* Apply for ECC Full Single setting */
+#define NSEC_DDR_0_BASE		0x47E00000U
+#define NSEC_DDR_0_SIZE		0x78200000
+#define NSEC_DDR_1_BASE		0x480000000U
+#define NSEC_DDR_1_SIZE		0x18000000
+#else
+/* Apply for non-ECC or Partial ECC setting */
+#define NSEC_DDR_0_BASE		0x47E00000U
+#define NSEC_DDR_0_SIZE		0x78200000
+#define NSEC_DDR_1_BASE		0x480000000U
+#define NSEC_DDR_1_SIZE		0x80000000
+#endif
+
+#else
+#error "Unknown platform flavor"
+#endif
+
+#endif /* RZG_COMMON_H */
diff --git a/core/arch/arm/plat-rzg/sub.mk b/core/arch/arm/plat-rzg/sub.mk
index b194244..796befd 100644
--- a/core/arch/arm/plat-rzg/sub.mk
+++ b/core/arch/arm/plat-rzg/sub.mk
@@ -2,4 +2,4 @@ global-incdirs-y += .
 srcs-y += main.c
 srcs-y += tee_common_otp.c
 
-cppflags-y += -DLTC_NO_PROTOTYPES
+cppflags-y += -DLTC_NO_PROTOTYPES -DRZG_DRAM_ECC=$(RZG_DRAM_ECC) -DRZG_ECC_FULL=$(RZG_ECC_FULL)
-- 
2.7.4

